### PR TITLE
Added plotting utilities and refactored hyperparams

### PR DESCRIPTION
**Pull request by ThibeauWouters**
This PR adds a few basic functionalities to simplify the treatment of the hyperparameters (now moved to a separate file) and by adding a few functionalities that allow to easily plot a few key quantities such as acceptance rates by a simple function call from any flowMC or jim script.

### CODE REVIEW

File: src/flowMC/sampler/Sampler.py [hunk 1]
```diff
@@ -7,36 +7,18 @@
...
 import numpy as np
-
+import matplotlib.pyplot as plt
+from flowMC.utils.hyperparameters import flowmc_default_hyperparameters
 
 class Sampler:
     """
     Sampler class that host configuration parameters, NF model, and local sampler
 
-    Args:
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:19]: All hyper-parameters doc strings should remain where it is for the doc to render correctly. Also instead of referring to more files, I rather have the parameters where they are called

File: src/flowMC/sampler/Sampler.py [hunk 2]
```diff
@@ -60,25 +42,14 @@ def __init__(
...
-        self.use_global = kwargs.get("use_global", True)
-        self.logging = kwargs.get("logging", True)
-        self.keep_quantile = kwargs.get("keep_quantile", 0)
-        self.local_autotune = kwargs.get("local_autotune", None)
-        self.train_thinning = kwargs.get("train_thinning", 1)
-        self.output_thinning = kwargs.get("output_thinning", 1)
-        self.n_sample_max = kwargs.get("n_sample_max", 10000)
-        self.precompile = kwargs.get("precompile", False)
-        self.verbose = kwargs.get("verbose", False)
+        # Set and override any given hyperparameters
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:45]: Thanks for the cleaning up here. In addition to this, instead of making a new `hyperparameters.py` file, I would keep the hyperparameters in the same File.

File: src/flowMC/sampler/Sampler.py [hunk 3]
```diff
@@ -98,19 +69,10 @@ def __init__(
         self.optim_state = tx.init(eqx.filter(self.nf_model, eqx.is_array))
         self.nf_training_loop, train_epoch, train_step = make_training_loop(tx)
 
-        # Initialized result dictionary
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:101]: Since this code is pretty small anyway, I don't see the point of adding a new file and a new function for this. Revert this part.

File: src/flowMC/sampler/Sampler.py [hunk 4]
```diff
@@ -495,3 +454,48 @@ def save_summary(self, path: str):
         """
         with open(path, "wb") as f:
             pickle.dump(self.summary, f)
+
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:457]: It was an explicit design choice to not include any plotting in the main sampler. We can look into adding a module for analysis and postprocessing in utils.

File: src/flowMC/utils/hyperparameters.py
Comment (kazewong) [inline src/flowMC/utils/hyperparameters.py:1]: See comments above related to hyper parameters

File: src/flowMC/utils/__init__.py
Comment (kazewong) [inline src/flowMC/utils/__init__.py:1]: See comments above regarding this extra init function

File: src/flowMC/sampler/Sampler.py (hunk 3 above)
Comment (ThibeauWouters) [reply src/flowMC/sampler/Sampler.py:101]: The idea was to reduce code duplication since it is used for both the training and production summaries. I have moved the function to the Sampler file, please let me know if you prefer to revert that as well.

File: src/flowMC/sampler/Sampler.py (hunk 4 above)
Comment (ThibeauWouters) [reply src/flowMC/sampler/Sampler.py:500]: I have moved these to a postprocessing file in utils.

File: src/flowMC/sampler/Sampler.py (hunk 3 above)
Comment (kazewong) [reply src/flowMC/sampler/Sampler.py:101]: Since the code is quite short here and it is only used twice (and intended to be only be used twice), I think having it being explicit like it is currently is preferred. Also, in general I would avoid string as a input for bookkeeping function, they can lead to all kinds of instability

File: src/flowMC/sampler/Sampler.py [hunk 5]
```diff
@@ -10,33 +10,58 @@
 from tqdm import tqdm
 import equinox as eqx
 import numpy as np
-
+import matplotlib.pyplot as plt
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:13]: Remove this unnecessary import

File: src/flowMC/sampler/Sampler.py (hunk 3 above)
Comment (ThibeauWouters) [reply src/flowMC/sampler/Sampler.py]: Ok, thanks for the feedback and insights into the code design! Reverted this change.

**Comment by kazewong** [comment]
@ThibeauWouters The checks seem to have failed due to some missing hyperparameters, would you mind have a look of that?

**Comment by ThibeauWouters** [comment]
@kazewong I have pushed some updates for the failed checks, and have also added some additional postprocessing analysis code to check the Gelman-Rubin R statistic.

**Comment by ThibeauWouters** [comment]
My tests show something is still off, I need to double check things

**Comment by ThibeauWouters** [comment]
It should be fixed now

File: src/flowMC/sampler/Sampler.py [hunk 6]
```diff
@@ -113,11 +128,16 @@ def __init__(
         production["log_prob"] = jnp.empty((self.n_chains, 0))
         production["local_accs"] = jnp.empty((self.n_chains, 0))
         production["global_accs"] = jnp.empty((self.n_chains, 0))
+        
+        if self.track_gelman_rubin:
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:132]: Since R_hat can be compute in post-processing, I don't think we should have it in the main sampler.

File: src/flowMC/sampler/Sampler.py [hunk 7]
```diff
@@ -264,6 +284,15 @@ def sampling_loop(
                 global_acceptance[:, 1::self.output_thinning],
                 axis=1,
             )
+            
+        if self.track_gelman_rubin:
```
Comment (kazewong) [inline src/flowMC/sampler/Sampler.py:288]: Same as above. Please remove this from the main sampler

File: src/flowMC/utils/postprocessing.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    plt.figure(figsize=figsize)
+    plt.plot(x, mean, linestyle="-", color="blue", alpha=alpha)
+    plt.xlabel("Iteration")
+    plt.ylabel(f"{name} ({which})")
+    # Extras for some variables:
+    if "acc" in name:
+        plt.ylim(0-eps, 1+eps)
+    plt.savefig(f"{outdir}{name}_{which}.png", bbox_inches='tight')
+    
+def gelman_rubin(chains: Float[Array, "n_chains n_steps n_dim"], discard_fraction: float = 0.1) -> Array:
```
Comment (kazewong) [inline src/flowMC/utils/postprocessing.py:63]: Arviz actually support r_hat and many more different statistical metric. I would suggest to use that instead of writing our own function.

**Review by kazewong — CHANGES_REQUESTED** [review]
I would not include r_hat related functionality in this PR. I am more inclined to have that as a notebook in example, since statistical metric like r_hat or correlation length needs to be used with context of user understanding the meaning of what they are measuring or monitoring.

**Review by kazewong — APPROVED** [review]
I took away the gelman rubin part of the previous PR.  Will merge it if the test does not fail
